src/analysis/destruct.ml

```diff
@@ -283,19 +283,42 @@ let collect_every_pattern_for_expression parent =
   ) Env.empty parent Location.none
   in loc, patterns
 
-
-
-let rec get_every_pattern = function
+let collect_function_pattern loc param_pattern =
+  match param_pattern.Typedtree.fp_kind with
+  | Typedtree.Tparam_pat pattern ->
+    loc, [pattern]
+  | Typedtree.Tparam_optional_default _ ->
+    raise (Not_allowed "value_binding")
+
+let rec get_every_pattern loc = function
   | [] -> assert false
   | parent :: parents ->
     match parent with
     | Case _
     | Pattern _ ->
       (* We are still in the same branch, going up. *)
-      get_every_pattern parents
+      get_every_pattern loc parents
     | Expression { exp_desc = Typedtree.Texp_ident (Path.Pident id, _, _) ; _}
       when Ident.name id = "*type-error*" ->
       raise (Ill_typed)
+    | Expression { exp_desc = Typedtree.Texp_function (params, _body); _ } ->
+      (* So we need to deal with the case where we're either in the body of a
+         function, or in a function parameter. *)
+      begin
+        match
+          List.find_some ~f:(fun param ->
+            let open Location in
+            let param_loc = param.Typedtree.fp_loc in
+            Lexing.compare_pos loc.loc_start param_loc.loc_start >= 0
+            && Lexing.compare_pos param_loc.loc_end loc.loc_end >= 0
```

Comment:
I think we should make a utility of that check that one location is included in another.
It looks like it does not exist yet. Maybe `Location_aux` is a good place to put it.
I suspect there are other usages of this in the codebase :-)

Reply:
Done in https://github.com/voodoos/merlin/pull/10/commits/df12650974583c22c5eea57fd48bc6f3b0fc7377 (not sure about the naming, WDYT?)

Reply:
I think `included` is the right keyword for this indeed.

Maybe we could use a labelled argument to actually show which of the two argument should be included in which ?

`Location_aux.is_included loc1 ~in:loc2`

Reply:
I go for `Location_aux.included ~into:parent loc` (`in` cannot be a label name). is that ok for you?